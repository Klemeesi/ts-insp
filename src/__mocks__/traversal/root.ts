import * as s1 from "./folder/submodule1";
import type * as s2 from "./folder/submodule2";
