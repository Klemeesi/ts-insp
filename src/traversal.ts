import * as ts from "typescript";
import * as fs from "fs";
import { ImportInfo, MainOptions, TraversalResult } from "./types";
import { getImportsFromNode } from "./helpers/importParser";

const isNodeModule = (absolutePath: string) => absolutePath.search("node_modules") != -1;

const getImportsFromFile = (options: MainOptions, filePath: string, currentLevel: number): ImportInfo[] => {
    const result: ImportInfo[] = [];
    const fileContent = fs.readFileSync(filePath, "utf-8");
    const sourceFile = ts.createSourceFile(filePath, fileContent, ts.ScriptTarget.Latest, true);

    options.logger("Traversing file:", filePath);

    ts.forEachChild(sourceFile, (childNode) => {
        // Run plugins
        options.inspOptions.plugins.forEach((plugin) => {
            options.logger(`Running ${plugin.name} plugin for`, filePath);
            plugin.processor(childNode, sourceFile, filePath, options);
        });

        const importInfos = getImportsFromNode(options, filePath, childNode, sourceFile);

        importInfos.forEach((importInfo) => {
            // Child node has an import
            options.logger("Found import:", importInfo.normalizedPath, `(${importInfo.absolutePath})`);
            result.push({
                import: importInfo.absolutePath || importInfo.normalizedPath,
                resolved: !!importInfo.absolutePath,
                absolutePath: importInfo.absolutePath,
                level: currentLevel,
                imports: [],
            });
        });
    });
    return result;
};

// Limit the number of iterations based on options
// Only process import if absolutePath is known. For some modules we don't know the absolutePath (like "fs")
// Limit node_modules traversal based on options
const shouldTraverse = (options: MainOptions, info: ImportInfo, importCounts: Record<string, number>) =>
    info.level < options.inspOptions.iterations &&
    info.absolutePath &&
    (options.inspOptions.traverseNodeModules || !isNodeModule(info.absolutePath)) &&
    (options.inspOptions.retraverse || !importCounts[info.import]);

// Second try with the function. Original was recursive and pretty complex method to modify.
// This one is supposed to be easier to read and not be recursive.
export const getImports = (options: MainOptions, filePath: string): TraversalResult => {
    const result: TraversalResult = { imports: [] };
    const importCounts: Record<string, number> = {};
    let nextBatchToProcess = [
        {
            parent: result,
            imports: getImportsFromFile(options, filePath, 1),
        },
    ];

    while (nextBatchToProcess.length) {
        const batch = [...nextBatchToProcess];
        nextBatchToProcess = [];

        batch.forEach((oneSet) => {
            const { parent } = oneSet;
            // Update parent imports. Mutates the object
            parent.imports = [...parent.imports, ...oneSet.imports];
            // Go thru each import and traverse them
            oneSet.imports.forEach((i) => {
                if (shouldTraverse(options, i, importCounts)) {
                    // If the import should be traversed, add it to the next batch
                    nextBatchToProcess = [
                        ...nextBatchToProcess,
                        {
                            parent: i,
                            imports: getImportsFromFile(options, i.absolutePath!, i.level + 1),
                        },
                    ];
                }
            });
            // Keep track of the import counts. Important if user has enabled retraversal
            oneSet.imports.forEach((i) => {
                importCounts[i.import] = importCounts[i.import] ? importCounts[i.import] + 1 : 1;
            });
        });
    }

    Object.keys(importCounts).forEach((key) => options.logger(key, "imported", importCounts[key], "times"));

    return result;
};
