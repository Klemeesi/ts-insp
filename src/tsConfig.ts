import * as ts from "typescript";
import * as fs from "fs";
import * as path from "path";

const getTsConfigFilePath = (initialFilePath: string) => {
  let currentDir = path.dirname(initialFilePath);
  while (currentDir !== "/") {
    const tsConfigFilePath = path.join(currentDir, "tsconfig.json");
    if (fs.existsSync(tsConfigFilePath)) {
      return { tsConfigFilePath, tsConfigPath: currentDir };
    }
    currentDir = path.dirname(currentDir);
  }
  return {};
};

export const getCompilerOptions = (initialFilePath: string) => {
  const { tsConfigFilePath, tsConfigPath } =
    getTsConfigFilePath(initialFilePath);
  if (!tsConfigFilePath) {
    return undefined;
  }

  const configFileContent = fs.readFileSync(tsConfigFilePath, "utf-8");
  const parsedConfig = ts.parseConfigFileTextToJson(
    tsConfigFilePath,
    configFileContent
  );

  if (parsedConfig.error) {
    console.error(
      "Error parsing tsconfig.json:",
      parsedConfig.error.messageText
    );
    return;
  }

  const config = ts.parseJsonConfigFileContent(
    parsedConfig.config,
    ts.sys,
    path.dirname(tsConfigFilePath)
  );
  return {
    tsConfigFilePath,
    tsConfigPath,
    options: config.options,
  };
};
