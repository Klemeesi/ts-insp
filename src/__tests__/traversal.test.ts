import { getImports } from "../traversal";
import { getCompilerOptions } from "../tsConfig";
import type { ImportInfoV2 } from "../types";

jest.mock("uuid", () => ({
    v4: jest.fn().mockReturnValue("mocked-uuid"), // Return a fixed ID 'mocked-uuid'
}));

const testConfig = {
    inspOptions: {
        file: "./src/__mocks__/traversal/root.ts",
        configFile: "",
        supportedTypes: ["ts", "js", "d.ts"],
        iterations: 10,
        verbose: false,
        retraverse: false,
        format: [],
    },
    logger: () => {},
    compilerOptions: getCompilerOptions("./tsconfig.json")?.options,
};

describe("traversal", () => {
    it("Basic traversal snapshot", () => {
        const result = getImports({ ...testConfig, compilerOptions: undefined }, testConfig.inspOptions.file);
        expect(result.imports).toMatchSnapshot();
    });

    it("Traversal with compiler options", () => {
        const result = getImports(testConfig, testConfig.inspOptions.file);
        expect(result.imports[0].moduleName).toBe("submodule1");
        expect(result.imports[0].imports[0].moduleName).toBe("fs");
        expect(result.imports[1].moduleName).toBe("submodule2");
        expect(result.imports[1].imports[0].moduleName).toBe("typescript");
        expect(result.imports[1].imports[1].moduleName).toBe("submodule1");
    });

    it("Traversal with retraverse", () => {
        const config = { ...testConfig, inspOptions: { ...testConfig.inspOptions, retraverse: true } };
        const result = getImports(config, testConfig.inspOptions.file);
        expect(result.imports[0].moduleName).toBe("submodule1");
        expect(result.imports[0].imports[0].moduleName).toBe("fs");
        expect(result.imports[1].moduleName).toBe("submodule2");
        expect(result.imports[1].imports[0].moduleName).toBe("typescript");
        expect(result.imports[1].imports[1].moduleName).toBe("submodule1");
        expect(result.imports[1].imports[1].imports[0].moduleName).toBe("fs");
    });

    it("Traversal with traverseNodeModules", () => {
        const config = { ...testConfig, inspOptions: { ...testConfig.inspOptions, traverseNodeModules: true } };
        const result = getImports(config, "./src/__mocks__/traversal/folder/submodule3.ts");
        expect(result.imports[0].moduleName).toBe("node-html-to-image");
        expect(result.imports[0].imports.length).toBeGreaterThan(0);
    });

    it("traversal with verbose", () => {
        const mockLog = jest.fn();
        const config = { ...testConfig, inspOptions: { ...testConfig.inspOptions, verbose: true }, logger: mockLog };
        getImports(config, testConfig.inspOptions.file);
        expect(mockLog).toHaveBeenCalled();
    });

    it("traversal with 1 iteration", () => {
        const config = { ...testConfig, inspOptions: { ...testConfig.inspOptions, iterations: 1 } };
        const result = getImports(config, testConfig.inspOptions.file);
        expect(result.imports[0].moduleName).toBe("submodule1");
        expect(result.imports[0].imports.length).toBe(0);
        expect(result.imports[1].moduleName).toBe("submodule2");
        expect(result.imports[1].imports.length).toBe(0);
    });

    it("traversal with supportedTypes", () => {
        const config = { ...testConfig, inspOptions: { ...testConfig.inspOptions, supportedTypes: ["ts"] } };
        const result = getImports(config, testConfig.inspOptions.file);
        expect(result.imports.length).toBe(1);
        expect(result.imports[0].moduleName).toBe("submodule1");
        expect(result.imports[0].imports.length).toBe(1);
        expect(result.imports[0].imports[0].moduleName).toBe("fs");
    });

    it("traversal with filterModules", () => {
        const config = {
            ...testConfig,
            inspOptions: {
                ...testConfig.inspOptions,
                filterModules: (a: ImportInfoV2) => {
                    return !a.absolutePath?.includes("node_modules");
                },
            },
        };
        const result = getImports(config, testConfig.inspOptions.file);
        expect(result.imports.length).toBe(2);
        expect(result.imports[0].moduleName).toBe("submodule1");
        expect(result.imports[0].imports.length).toBe(1);
        expect(result.imports[0].imports[0].moduleName).toBe("fs");
        expect(result.imports[1].moduleName).toBe("submodule2");
        expect(result.imports[1].imports.length).toBe(1);
        expect(result.imports[1].imports[0].moduleName).toBe("submodule1");
    });

    it("traversal with skipTypeImports", () => {
        const config = {
            ...testConfig,
            inspOptions: {
                ...testConfig.inspOptions,
                skipTypeImports: true,
            },
        };
        const result = getImports(config, testConfig.inspOptions.file);
        expect(result.imports.length).toBe(1);
        expect(result.imports[0].moduleName).toBe("submodule1");
        expect(result.imports[0].imports.length).toBe(1);
        expect(result.imports[0].imports[0].moduleName).toBe("fs");
    });
});
