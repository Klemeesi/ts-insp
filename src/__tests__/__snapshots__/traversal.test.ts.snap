// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traversal Basic traversal test 1`] = `
[
  {
    "absolutePath": "C:\\Source\\ts-insp\\src\\traversal.ts",
    "import": "C:\\Source\\ts-insp\\src\\traversal.ts",
    "imports": [
      {
        "import": "typescript",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "import": "path",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "import": "fs",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "import": "./types",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "absolutePath": "C:\\Source\\ts-insp\\src\\output\\log.ts",
        "import": "C:\\Source\\ts-insp\\src\\output\\log.ts",
        "imports": [],
        "level": 2,
        "resolved": true,
      },
    ],
    "level": 1,
    "resolved": true,
  },
  {
    "absolutePath": "C:\\Source\\ts-insp\\src\\tsConfig.ts",
    "import": "C:\\Source\\ts-insp\\src\\tsConfig.ts",
    "imports": [
      {
        "import": "typescript",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "import": "fs",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "import": "path",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
    ],
    "level": 1,
    "resolved": true,
  },
]
`;

exports[`traversal Traversal test with compiler options 1`] = `
[
  {
    "absolutePath": "src/traversal.ts",
    "import": "src/traversal.ts",
    "imports": [
      {
        "absolutePath": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
        "import": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
        "imports": [],
        "level": 2,
        "resolved": true,
      },
      {
        "import": "path",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "import": "fs",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "absolutePath": "src/types.d.ts",
        "import": "src/types.d.ts",
        "imports": [
          {
            "absolutePath": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
            "import": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
            "imports": [],
            "level": 3,
            "resolved": true,
          },
          {
            "absolutePath": "src/plugins.ts",
            "import": "src/plugins.ts",
            "imports": [
              {
                "absolutePath": "src/plugins/debugPlugin.ts",
                "import": "src/plugins/debugPlugin.ts",
                "imports": [
                  {
                    "absolutePath": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                    "import": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                    "imports": [],
                    "level": 5,
                    "resolved": true,
                  },
                  {
                    "absolutePath": "src/types.d.ts",
                    "import": "src/types.d.ts",
                    "imports": [
                      {
                        "absolutePath": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                        "import": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                        "imports": [],
                        "level": 6,
                        "resolved": true,
                      },
                      {
                        "absolutePath": "src/plugins.ts",
                        "import": "src/plugins.ts",
                        "imports": [
                          {
                            "absolutePath": "src/plugins/debugPlugin.ts",
                            "import": "src/plugins/debugPlugin.ts",
                            "imports": [
                              {
                                "absolutePath": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                                "import": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                                "imports": [],
                                "level": 8,
                                "resolved": true,
                              },
                              {
                                "absolutePath": "src/types.d.ts",
                                "import": "src/types.d.ts",
                                "imports": [
                                  {
                                    "absolutePath": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                                    "import": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
                                    "imports": [],
                                    "level": 9,
                                    "resolved": true,
                                  },
                                  {
                                    "absolutePath": "src/plugins.ts",
                                    "import": "src/plugins.ts",
                                    "imports": [
                                      {
                                        "absolutePath": "src/plugins/debugPlugin.ts",
                                        "import": "src/plugins/debugPlugin.ts",
                                        "imports": [],
                                        "level": 10,
                                        "resolved": true,
                                      },
                                    ],
                                    "level": 9,
                                    "resolved": true,
                                  },
                                ],
                                "level": 8,
                                "resolved": true,
                              },
                            ],
                            "level": 7,
                            "resolved": true,
                          },
                        ],
                        "level": 6,
                        "resolved": true,
                      },
                    ],
                    "level": 5,
                    "resolved": true,
                  },
                ],
                "level": 4,
                "resolved": true,
              },
            ],
            "level": 3,
            "resolved": true,
          },
        ],
        "level": 2,
        "resolved": true,
      },
      {
        "absolutePath": "src/output/log.ts",
        "import": "src/output/log.ts",
        "imports": [],
        "level": 2,
        "resolved": true,
      },
    ],
    "level": 1,
    "resolved": true,
  },
  {
    "absolutePath": "src/tsConfig.ts",
    "import": "src/tsConfig.ts",
    "imports": [
      {
        "absolutePath": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
        "import": "C:/Source/ts-insp/node_modules/typescript/lib/typescript.d.ts",
        "imports": [],
        "level": 2,
        "resolved": true,
      },
      {
        "import": "fs",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
      {
        "import": "path",
        "imports": [],
        "level": 2,
        "resolved": false,
      },
    ],
    "level": 1,
    "resolved": true,
  },
]
`;
